[CMake] Detect BoringSSL

donny-dont opened the pull request:
#### f5806aff636facd0cb44818dbbb9d3f9868adaa1
<pre>
[CMake] Detect BoringSSL
<a href="https://bugs.webkit.org/show_bug.cgi?id=244050">https://bugs.webkit.org/show_bug.cgi?id=244050</a>

Reviewed by Michael Catanzaro.

Determine if the OpenSSL implementation is BoringSSL by looking for the symbol
`OPENSSL_IS_BORINGSSL`. This information is needed to determine which WebCrypto algorithms can be
implemented. The block of USE declarations specific to WebCore is now checked with `ENABLE_WEBCORE`
since FindOpenSSL is only used when that is enabled on PlayStation.

* Source/cmake/OptionsPlayStation.cmake:
* Source/cmake/OptionsWinCairo.cmake:
* Source/cmake/WebKitCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/253546@main">https://commits.webkit.org/253546@main</a>
</pre>











<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5993186c5beca3be78795fe4bcbb7653006125e9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86320 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30241 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95168 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28604 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78456 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90424 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/91919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23201 "Passed tests") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23302 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66308 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78268 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26563 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/71890 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26472 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13509 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25721 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2526 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28148 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/74672 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/32790 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16516 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/cmake/OptionsWinCairo.cmake:
```diff
@@ -67,6 +67,13 @@ SET_AND_EXPOSE_TO_BUILD(ENABLE_DEVELOPER_MODE ${DEVELOPER_MODE})
+cmake_push_check_state()
+set(CMAKE_REQUIRED_INCLUDES "${OPENSSL_INCLUDE_DIR}")
+set(CMAKE_REQUIRED_LIBRARIES "${OPENSSL_LIBRARIES}")
+WEBKIT_CHECK_HAVE_SYMBOL(USE_BORINGSSL OPENSSL_IS_BORINGSSL openssl/ssl.h)
```
If BoringSSL has `OPENSSL_IS_BORINGSSL` macro, do we have to define `USE_BORINGSSL`? How about an idea using `defined(OPENSSL_IS_BORINGSSL)`?

donny-dont replied:
Combing with the question above yes some algorithm's are not present in BoringSSL. The one that is going to have to be gated on whether or not its BoringSSL is `CryptoAlgorithmAES_CFB`.

With regard to using `OPENSSL_IS_BORINGSSL` directly it was more to fit with the conventions in WebKit and to allow us to control the value inside the build rather than relying on the outside one.

donny-dont replied:
You can see what algorithms are present by looking at https://diafygi.github.io/webcrypto-examples/

It looks like AES-CFB is actually no longer in the spec

donny-dont committed:
[CMake] Detect BoringSSL
https://bugs.webkit.org/show_bug.cgi?id=244050

Reviewed by Michael Catanzaro.

Determine if the OpenSSL implementation is BoringSSL by looking for the symbol
`OPENSSL_IS_BORINGSSL`. This information is needed to determine which WebCrypto algorithms can be
implemented. The block of USE declarations specific to WebCore is now checked with `ENABLE_WEBCORE`
since FindOpenSSL is only used when that is enabled on PlayStation.

* Source/cmake/OptionsPlayStation.cmake:
* Source/cmake/OptionsWinCairo.cmake:
* Source/cmake/WebKitCommon.cmake:

Canonical link: https://commits.webkit.org/253546@main
